h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -678,114 +697,311 @@ public DataType getKeyType() {
         return map.getKeyType();
     }
 
+    private <X> TMIterator<K, X> getRepeatableIterator(K from, K to, boolean forEntries) {
+        return transaction.hasChanges() ? new RepeatableIterator<K, X>(this, from, to, forEntries)
+                : new CommittedIterator<K, X>(this, from, to, forEntries);
+    }
+
+    private static final class UncommittedIterator<K, X> extends TMIterator<K, X> {
 
-    private static final class KeyIterator<K> extends TMIterator<K,K> {
+        private final int transactionId;
+
+        private final Cursor<K, VersionedValue> cursor;
 
-        public KeyIterator(TransactionMap<K, ?> transactionMap, K from, K to, boolean includeUncommitted) {
-            super(transactionMap, from, to, includeUncommitted);
+        UncommittedIterator(TransactionMap<K, ?> transactionMap, K from, K to, boolean forEntries) {
+            super(forEntries);
+            Transaction transaction = transactionMap.getTransaction();
+            this.transactionId = transaction.transactionId;
+            this.cursor = new Cursor<>(transactionMap.map.flushAndGetRoot().root, from, to);
+            fetchNext();
         }
 
         @Override
-        protected K registerCurrent(K key, VersionedValue data) {
-            return key;
+        void fetchNext() {
+            while (cursor.hasNext()) {
+                K key = cursor.next();
+                VersionedValue data = cursor.getValue();
+                if (data != null) {
+                    Object currentValue = data.getCurrentValue();
+                    if (currentValue != null
+                            || transactionId != TransactionStore.getTransactionId(data.getOperationId())) {
```

Comment:
I think, I got it wrong here, and you just made it more obvious after split.
If some other transaction deleted an entry, this condition will be true, but that might be wrong, if that transaction already committed (committing), but is not closed yet.
That's why committingTransactions bit mask has to be consulted here as well.

Reply:
I can add this mask to the iterator, but how exactly the correct check should look like?

Reply:
Similar to another iterator:
```
        long id;
        int tx;
        if (currentValue != null
                || (id = data.getOperationId()) != 0 
                    && transactionId != (tx = TransactionStore.getTransactionId(id))
                    && !committingTransactions.get(tx)) {
```
or more verbose but straightforward way
```
    if (currentValue != null) {
        registerCurrent(key, currentValue);
        return;
    }
    long id = data.getOperationId();
    if (id != 0) {
        int tx = TransactionStore.getTransactionId(id);
        if (transactionId != tx && !committingTransactions.get(tx)) {
            registerCurrent(key, currentValue);
            return;
        }
    }

```

Reply:
I used the verbose version, thanks.